### ISSUE TITLE
About DOC's .h5 data

### ISSUE THREAD

**Issue by buptlr** [GuoxiaWang/DOOBNet#1]
Hi GUOxia Wang！recently，I tried to train the ori net of DOC，but it needs .h5 format，so I  use your "doobnet_mat2hdf5_edge_ori.py" to tramsform .mat to .h5，and then I get the .h5 training dataset which includes the edge map and ori map channels. But when I train the ori net I encourntered the INF and NAN error in loss. I wonder whether the .h5 dataset can directly apply to DOC or maybe I should make a modification.Thanks for your time!

**Comment by GuoxiaWang** [comment]
@buptlr 

Yes, you are right.

You need to modify DOC's train_val_doc_ori.prototxt. Focus the following layers:
```
layer {
    name: "slice"
    type: "Slice"
    bottom: "label"
    top: 'label_edge'
    top: 'res_edge'
    top: 'label_occ'
    slice_param{
        slice_point: 1
        slice_point: 2
    }
}
layer { type: "Silence" bottom: "res_edge" } 
```
The second channel of DOC's .h5 is not used, maybe is used to the other task.

And doobnet_mat2hdf5_edge_ori.py coverts .mat to .h5 which has two channels, the first channel is edge map and the second is ori map.

**Comment by buptlr** [comment]
@GuoxiaWang 
Thanks for your answer! However, I have commented the res_edge, slice point 2 and layer Silence in train_val_doc_ori.prototxt and the loss is still INF. I have no idea why the loss is not correct. Could I bother you to help me check the files to train at your convenience?it may take up your time and thank you a lot in advance!

The solver.py is almost the same with hed，I only modify the sys.path.append and caffe.SGDSolver and I wonder whether base_weights shouldn't be the '5stage-vgg.caffemodel'?

    from __future__ import division
    import sys
    sys.path.append('/home/xuefeng/project/DOC-master/python/')
    import caffe
    import numpy as np

    def upsample_filt(size):
        factor = (size + 1) // 2
    if size % 2 == 1:
        center = factor - 1
    else:
        center = factor - 0.5
    og = np.ogrid[:size, :size]
    return (1 - abs(og[0] - center) / factor) * \
           (1 - abs(og[1] - center) / factor)

    def interp_surgery(net, layers):
        for l in layers:
            m, k, h, w = net.params[l][0].data.shape
            if m != k:
                print 'input + output channels need to be the same'
                raise
            if h != w:
                print 'filters need to be square'
                raise
            filt = upsample_filt(h)
            net.params[l][0].data[range(m), range(k), :, :] = filt

    base_weights = '5stage-vgg.caffemodel'

    caffe.set_mode_gpu()
    caffe.set_device(1)

    solver = caffe.SGDSolver('solver_doc_ori.prototxt')

    interp_layers = [k for k in solver.net.params.keys() if 'up' in k]
    interp_surgery(solver.net, interp_layers)

    solver.net.copy_from(base_weights)
    solver.step(30000)

The solver_doc_ori.prototxt is not changed

    net: "train_val_doc_ori.prototxt"
    test_iter: 0
    test_interval: 1000000
    base_lr: 0.0001
    lr_policy: "step"
    gamma: 0.1
    iter_size: 10
    stepsize: 5000
    display: 5
    max_iter: 50000
    momentum: 0.9
    weight_decay: 0.0005
    snapshot: 2500
    snapshot_prefix: "./models/ori_hed"

I changed train_val_doc_ori.prototxt in these: comment the two lines in slice layer and comment the slience layer.

    name: "HED"
    layer {
        name: "data"
        type: "ImageLabelmapData"
        top: "data"
        top: "label"
        include {
            phase: TRAIN
        }
        transform_param {
        mirror: false
        mean_value: 104.00699
        mean_value: 116.66877
        mean_value: 122.67892
        }
        image_data_param {
        root_folder: "../DOC-master/data/aug/"
        source: "../DOC-master/data/aug/train_pair.lst"
        data_type: "h5"
        batch_size: 1
        shuffle: true
        new_height: 0
        new_width: 0
        }
    }

    layer {
        name: "data"
        type: "ImageLabelmapData"
        top: "data"
        top: "label"
        include {
            phase: TEST
        }
        transform_param {
            mirror: false
            mean_value: 104.00699
            mean_value: 116.66877
            mean_value: 122.67892
        }
        image_data_param {
        root_folder: "../DOC-master/data/aug/"
        source: "../DOC-master/data/aug/train_pair.lst"
        data_type: "h5"
        batch_size: 1 # only set for one if using adaptive size 
        shuffle: true
        new_height: 0
        new_width: 0
        }
    }

    layer {
	    name: "slice"
   	    type: "Slice"
   	    bottom: "label"
   	    top: 'label_edge'
            #top: 'res_edge'
	    top: 'label_occ'
	    slice_param{
	    slice_point: 1
            #slice_point: 2
	    }
    }
    #layer { type: "Silence" bottom: "res_edge" } 

And the Command Line result is as follows, the loss is either nan or inf at the beginning of the train phase and the diffs are almost 0:

    Iteration 20, loss = -nan
    dsn3_loss_occ = inf (* 1 = inf loss)
    Train net output #1: dsn4_loss_occ = inf (* 1 = inf loss)
    Train net output #2: dsn5_loss_occ = inf (* 1 = inf loss)
    Train net output #3: fuse_loss_occ = inf (* 1 = inf loss)
    Iteration 20, lr = 0.0001
    [Update] Layer conv1_1, param 0 data: 0.168639; diff: 7.50935e-08
    [Update] Layer conv1_1, param 1 data: 0.501886; diff: 0
    [Update] Layer conv1_2, param 0 data: 0.0292294; diff: 1.30156e-08
    [Update] Layer conv1_2, param 1 data: 0.265496; diff: 0
    [Update] Layer conv2_1, param 0 data: 0.0208213; diff: 9.27154e-09
    [Update] Layer conv2_1, param 1 data: 0.133923; diff: 0
    [Update] Layer conv2_2, param 0 data: 0.0158208; diff: 7.04487e-09
    [Update] Layer conv2_2, param 1 data: 0.147183; diff: 0
    [Update] Layer conv3_1, param 0 data: 0.0119275; diff: 5.31122e-09
    [Update] Layer conv3_1, param 1 data: 0.054964; diff: 1.64474e-19
    [Update] Layer conv3_2, param 0 data: 0.00887878; diff: 3.95365e-09
    [Update] Layer conv3_2, param 1 data: 0.0632454; diff: 5.06844e-20
    [Update] Layer conv3_3, param 0 data: 0.00905534; diff: 4.03227e-09
    [Update] Layer conv3_3, param 1 data: 0.0591653; diff: 1.92947e-20
    [Update] Layer conv4_1, param 0 data: 0.00731396; diff: 3.25685e-09
    [Update] Layer conv4_1, param 1 data: 0.0411595; diff: 6.6842e-22
    [Update] Layer conv4_2, param 0 data: 0.00576387; diff: 2.5666e-09
    [Update] Layer conv4_2, param 1 data: 0.0425173; diff: 2.82529e-22
    [Update] Layer conv4_3, param 0 data: 0.00598295; diff: 2.66416e-09
    [Update] Layer conv4_3, param 1 data: 0.0568271; diff: 1.15697e-22
    [Update] Layer conv5_1, param 0 data: 0.00658367; diff: 2.93237e-07
    [Update] Layer conv5_1, param 1 data: 0.108178; diff: 5.21865e-19
    [Update] Layer conv5_2, param 0 data: 0.00679712; diff: 3.02744e-07
    [Update] Layer conv5_2, param 1 data: 0.165417; diff: 1.93814e-19
    [Update] Layer conv5_3, param 0 data: 0.00654682; diff: 2.91596e-07
    [Update] Layer conv5_3, param 1 data: 0.219785; diff: 3.26019e-20
    [Update] Layer score-dsn3-occ, param 0 data: 3.57067e-09; diff: 6.30604e-11
    [Update] Layer score-dsn3-occ, param 1 data: 1.71736e-08; diff: 3.03298e-10
    [Update] Layer upsample_4_occ, param 0 data: 0.25; diff: 0
    [Update] Layer upsample_4_occ, param 1 data: 0; diff: 0
    [Update] Layer score-dsn4-occ, param 0 data: 8.23372e-12; diff: 3.68535e-13
    [Update] Layer score-dsn4-occ, param 1 data: 7.56449e-11; diff: 3.38581e-12
    [Update] Layer upsample_8_occ, param 0 data: 0.25; diff: 0
    [Update] Layer upsample_8_occ, param 1 data: 0; diff: 0
    [Update] Layer score-dsn5, param 0 data: 1.45992e-10; diff: 2.59117e-12
    [Update] Layer score-dsn5, param 1 data: 2.53182e-10; diff: 4.49383e-12
    [Update] Layer upsample_16_occ, param 0 data: 0.25; diff: 0
    [Update] Layer upsample_16_occ, param 1 data: 0; diff: 0
    [Update] Layer new-score-weighting-occ, param 0 data: 0.3; diff: 1.33587e-08
    [Update] Layer new-score-weighting-occ, param 1 data: 2.63267e-12; diff: 6.68184e-14
    Data layer prefetch queue empty
    [Backward] All net params (data, diff): L1 norm = (103471, 7.66314e-09); L2 norm = (43.395, 3.48882e-10)
    [Backward] All net params (data, diff): L1 norm = (103471, 7.67259e-09); L2 norm = (43.395, 3.49311e-10)
    [Backward] All net params (data, diff): L1 norm = (103471, 9.15869e-05); L2 norm = (43.395, 7.0676e-06)
    [Backward] All net params (data, diff): L1 norm = (103471, 9.15869e-05); L2 norm = (43.395, 7.06758e-06)
    [Backward] All net params (data, diff): L1 norm = (103471, 9.15869e-05); L2 norm = (43.395, 7.06758e-06)
    [Backward] All net params (data, diff): L1 norm = (103471, 9.1596e-05); L2 norm = (43.395, 7.06689e-06)
    [Backward] All net params (data, diff): L1 norm = (103471, 9.1596e-05); L2 norm = (43.395, 7.06689e-06)
    [Backward] All net params (data, diff): L1 norm = (103471, 0.00132858); L2 norm = (43.395, 0.000113826)
    [Backward] All net params (data, diff): L1 norm = (103471, 0.00132858); L2 norm = (43.395, 0.000113826)
    [Backward] All net params (data, diff): L1 norm = (103471, 0.00132858); L2 norm = (43.395, 0.000113826)

**Comment by GuoxiaWang** [comment]
@buptlr 

I have tried to train train_val_doc_ori.prototxt.  Here are some logs:
```
I0622 13:03:42.827455  8114 sgd_solver.cpp:112] Iteration 0, lr = 0.0001
I0622 13:03:49.936727  8114 solver.cpp:239] Iteration 5 (0.703283 iter/s, 7.10951s/5 iters), loss = 9.9011
I0622 13:03:49.936800  8114 solver.cpp:258]     Train net output #0: dsn3_loss_occ = 1.54939 (* 1 = 1.54939 loss)
I0622 13:03:49.936816  8114 solver.cpp:258]     Train net output #1: dsn4_loss_occ = 2.82174 (* 1 = 2.82174 loss)
I0622 13:03:49.936827  8114 solver.cpp:258]     Train net output #2: dsn5_loss_occ = 3.35637 (* 1 = 3.35637 loss)
I0622 13:03:49.936851  8114 solver.cpp:258]     Train net output #3: fuse_loss_occ = 3.48447 (* 1 = 3.48447 loss)
I0622 13:03:49.936862  8114 sgd_solver.cpp:112] Iteration 5, lr = 0.0001
I0622 13:03:57.000193  8114 solver.cpp:239] Iteration 10 (0.707859 iter/s, 7.06355s/5 iters), loss = 11.5593
I0622 13:03:57.000267  8114 solver.cpp:258]     Train net output #0: dsn3_loss_occ = 2.61916 (* 1 = 2.61916 loss)
I0622 13:03:57.000284  8114 solver.cpp:258]     Train net output #1: dsn4_loss_occ = 1.76694 (* 1 = 1.76694 loss)
I0622 13:03:57.000295  8114 solver.cpp:258]     Train net output #2: dsn5_loss_occ = 2.2206 (* 1 = 2.2206 loss)
I0622 13:03:57.000320  8114 solver.cpp:258]     Train net output #3: fuse_loss_occ = 1.70487 (* 1 = 1.70487 loss)
I0622 13:03:57.000334  8114 sgd_solver.cpp:112] Iteration 10, lr = 0.0001
I0622 13:04:04.097358  8114 solver.cpp:239] Iteration 15 (0.704499 iter/s, 7.09724s/5 iters), loss = 12.0651
I0622 13:04:04.097419  8114 solver.cpp:258]     Train net output #0: dsn3_loss_occ = 4.81059 (* 1 = 4.81059 loss)
I0622 13:04:04.097438  8114 solver.cpp:258]     Train net output #1: dsn4_loss_occ = 1.6307 (* 1 = 1.6307 loss)
I0622 13:04:04.097451  8114 solver.cpp:258]     Train net output #2: dsn5_loss_occ = 2.02154 (* 1 = 2.02154 loss)
I0622 13:04:04.097463  8114 solver.cpp:258]     Train net output #3: fuse_loss_occ = 1.99214 (* 1 = 1.99214 loss)
I0622 13:04:04.097476  8114 sgd_solver.cpp:112] Iteration 15, lr = 0.0001
I0622 13:04:11.154549  8114 solver.cpp:239] Iteration 20 (0.708489 iter/s, 7.05728s/5 iters), loss = 7.63927
I0622 13:04:11.154618  8114 solver.cpp:258]     Train net output #0: dsn3_loss_occ = 3.21601 (* 1 = 3.21601 loss)
I0622 13:04:11.154633  8114 solver.cpp:258]     Train net output #1: dsn4_loss_occ = 1.30635 (* 1 = 1.30635 loss)
I0622 13:04:11.154645  8114 solver.cpp:258]     Train net output #2: dsn5_loss_occ = 1.54567 (* 1 = 1.54567 loss)
I0622 13:04:11.154670  8114 solver.cpp:258]     Train net output #3: fuse_loss_occ = 1.57694 (* 1 = 1.57694 loss)

```

All the setting I used above is the default in the DOC's protos folder and I use the .h5 data converting from doobnet_mat2hdf5_edge_ori.py, except to comment Slice and Silence layer and modify ImageLabelmapData layer's root_folder and source setting in train_val_doc_ori.prototxt.

Please check carefully your setting.

**Comment by buptlr** [comment]
Thanks for your reply and your e-mail, but I can't reply the e-mail due to the address error. I tried to reduce the base_lr to 0.00001 and 0.000001, but it still can't work. Now I know my training process is correct and maybe I changed something by accident. I am going to restart all the training processes and may it will work. 
Thanks again for your valuable time！
